packages/fuels-types/src/param_types.rs

```diff
@@ -1295,53 +1334,116 @@ mod tests {
             variants: EnumVariants::new(tuples_no_nested_vec)?,
             generics: param_types_nested_vec
         }
-        .contains_nested_vectors());
+        .contains_nested_heap_types());
+        Ok(())
+    }
 
+    #[test]
+    fn contains_no_nested_heap_type_nested_bytes_complex_types() -> Result<()> {
+        let base_bytes = ParamType::Bytes;
+        let tuples_no_nested_bytes = vec![
+            ("Bim".to_string(), ParamType::U16),
+            ("Bam".to_string(), ParamType::Bool),
+        ];
+        let tuples_with_nested_bytes = vec![
+            ("Zim".to_string(), ParamType::U64),
+            ("Zam".to_string(), ParamType::U32),
+            ("Boum".to_string(), base_bytes.clone()),
+        ];
+        let param_types_no_nested_bytes = vec![ParamType::U64, ParamType::U32];
+        let param_types_nested_bytes = vec![ParamType::Unit, ParamType::Bool, base_bytes.clone()];
+
+        assert!(!base_bytes.contains_nested_heap_types());
+        assert!(ParamType::Vector(Box::from(base_bytes.clone())).contains_nested_heap_types());
+
+        assert!(!ParamType::Array(Box::from(ParamType::U8), 10).contains_nested_heap_types());
+        assert!(ParamType::Array(Box::from(base_bytes), 10).contains_nested_heap_types());
+
+        assert!(!ParamType::Tuple(param_types_no_nested_bytes.clone()).contains_nested_heap_types());
+        assert!(ParamType::Tuple(param_types_nested_bytes.clone()).contains_nested_heap_types());
+
+        assert!(!ParamType::Struct {
+            name: "StructName".to_string(),
+            generics: param_types_no_nested_bytes.clone(),
+            fields: tuples_no_nested_bytes.clone(),
+        }
+        .contains_nested_heap_types());
+        assert!(ParamType::Struct {
+            name: "StructName".to_string(),
+            generics: param_types_nested_bytes.clone(),
+            fields: tuples_no_nested_bytes.clone()
+        }
+        .contains_nested_heap_types());
+        assert!(ParamType::Struct {
+            name: "StructName".to_string(),
+            generics: param_types_no_nested_bytes.clone(),
+            fields: tuples_with_nested_bytes.clone()
+        }
+        .contains_nested_heap_types());
+
+        assert!(!ParamType::Enum {
+            name: "EnumName".to_string(),
+            variants: EnumVariants::new(tuples_no_nested_bytes.clone())?,
+            generics: param_types_no_nested_bytes.clone()
+        }
+        .contains_nested_heap_types());
+        assert!(ParamType::Enum {
+            name: "EnumName".to_string(),
+            variants: EnumVariants::new(tuples_with_nested_bytes)?,
+            generics: param_types_no_nested_bytes
+        }
+        .contains_nested_heap_types());
+        assert!(ParamType::Enum {
+            name: "EnumName".to_string(),
+            variants: EnumVariants::new(tuples_no_nested_bytes)?,
+            generics: param_types_nested_bytes
+        }
+        .contains_nested_heap_types());
```

Comment:
You can use a helper fn to remove some of the noise around calling `assert!` and `contains_nested_heap_types` all the time:
```suggestion
        let is_nested = |param_type: ParamType| assert!(param_type.contains_nested_heap_types());
        let not_nested = |param_type: ParamType| assert!(!param_type.contains_nested_heap_types());

        not_nested(base_bytes.clone());
        is_nested(ParamType::Vector(Box::from(base_bytes.clone())));

        not_nested(ParamType::Array(Box::from(ParamType::U8), 10));
        is_nested(ParamType::Array(Box::from(base_bytes), 10));

        not_nested(ParamType::Tuple(param_types_no_nested_bytes.clone()));
        is_nested(ParamType::Tuple(param_types_nested_bytes.clone()));

        let not_nested_struct = ParamType::Struct {
            name: "StructName".to_string(),
            generics: param_types_no_nested_bytes.clone(),
            fields: tuples_no_nested_bytes.clone(),
        };
        not_nested(not_nested_struct);

        let nested_struct = ParamType::Struct {
            name: "StructName".to_string(),
            generics: param_types_nested_bytes.clone(),
            fields: tuples_no_nested_bytes.clone(),
        };
        is_nested(nested_struct);

        let nested_struct = ParamType::Struct {
            name: "StructName".to_string(),
            generics: param_types_no_nested_bytes.clone(),
            fields: tuples_with_nested_bytes.clone(),
        };
        is_nested(nested_struct);

        let not_nested_enum = ParamType::Enum {
            name: "EnumName".to_string(),
            variants: EnumVariants::new(tuples_no_nested_bytes.clone())?,
            generics: param_types_no_nested_bytes.clone(),
        };
        not_nested(not_nested_enum);

        let nested_enum = ParamType::Enum {
            name: "EnumName".to_string(),
            variants: EnumVariants::new(tuples_with_nested_bytes)?,
            generics: param_types_no_nested_bytes,
        };
        is_nested(nested_enum);

        let nested_enum = ParamType::Enum {
            name: "EnumName".to_string(),
            variants: EnumVariants::new(tuples_no_nested_bytes)?,
            generics: param_types_nested_bytes,
        };
        is_nested(nested_enum);
```